modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java

```diff
@@ -646,7 +646,11 @@ public void accept(RestChannel channel) {
                                 client.execute(TYPE, new Request(), new RestActionListener<>(channel) {
                                     @Override
                                     protected void processResponse(Response response) {
-                                        localRefs.mustIncRef();
+                                        // incRef can fail if the request was already cancelled
+                                        if (localRefs.tryIncRef() == false) {
+                                            assert localRefs.hasReferences() == false : "tryIncRef failed but RefCounted not completed";
+                                            return;
+                                        }
```

Comment:
This feels like it could hide the test being broken (i.e. if we always cancel before the call to `processResponse` it'll happily pass without testing anything.

Another approach would be to put a synchronization point after the call to sendResponse, which we wait on before cancelling the request. That way we could be sure we'd begun sending the response before the cancellation? but perhaps handling cancellation prior to the response being sent is part of the test?

Reply:
I guess we could just call `tryIncRef`, and if it fails to increment the counter, it should also not decrement it later (by virtue of `channel.sendResponse` not sending the chunked response)?

Reply:
I think the current change is fine. We also do something similar at `prepareRequest` a few lines above.

https://github.com/elastic/elasticsearch/blob/ef52f8ccb828a2e096b972e681fd53c2b3ab598b/modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java#L637

> perhaps handling cancellation prior to the response being sent is part of the test?

I think so. The random sleep in the test method is intended to have cancellation happen in different places, including in `prepareRequest` where no response is ever generated.

> it should also not decrement it later (by virtue of channel.sendResponse not sending the chunked response)?

I think this might not be the case. It seems that response is released if `channel.sendResponse` fails to send.

https://github.com/elastic/elasticsearch/blob/ef52f8ccb828a2e096b972e681fd53c2b3ab598b/server/src/main/java/org/elasticsearch/http/DefaultRestChannel.java#L191-L192